src/CircuitOptimizer.cpp

```diff
@@ -667,4 +667,79 @@ namespace qc {
+        //      └───┘└╥┘     └───┘└╥┘   -->      └───┘└╥┘ ░ ┌───┐┌─┐
+        // c: 2/══════╩════════════╩═       q_1: ──────╫──░─┤ H ├┤M├
+        //            0            1                   ║  ░ └───┘└╥┘
+        //                                  c: 2/══════╩══════════╩═
+        //                                             0          1
+        auto replacementMap = std::map<dd::Qubit, dd::Qubit>();
+        for (auto it = qc.ops.begin(); it != qc.ops.end(); it++) {
+            if (!replacementMap.empty()) {
+                if ((*it)->isStandardOperation()) {
+                    changeTargets(reinterpret_cast<Operation&>(*it), replacementMap);
+                    changeControls(reinterpret_cast<Operation&>(*it), replacementMap);
+                }
+                if ((*it)->isNonUnitaryOperation()) {
+                    changeTargets(reinterpret_cast<Operation&>(*it), replacementMap);
+                }
+                if ((*it)->isCompoundOperation()) {
+                    changeTargets(reinterpret_cast<Operation&>(*it), replacementMap);
+                    changeControls(reinterpret_cast<Operation&>(*it), replacementMap);
+                }
+                if ((*it)->isClassicControlledOperation()) {
+                    changeTargets(reinterpret_cast<Operation&>(*it), replacementMap);
+                    changeControls(reinterpret_cast<Operation&>(*it), replacementMap);
+                }
+            }
+            if ((*it)->getType() == qc::Reset) {
+                auto indexAddQubit = static_cast<dd::Qubit>(qc.getNqubits());
+                qc.addQubit(indexAddQubit, indexAddQubit, indexAddQubit);
+                for (const auto& target: (*it)->getTargets()) {
+                    replacementMap.insert(std::pair(static_cast<dd::Qubit>(target), static_cast<dd::Qubit>(indexAddQubit)));
+                }
+                it = qc.erase(it);
+            }
+        }
+    }
+
+    void CircuitOptimizer::changeTargets(Operation& op, std::map<dd::Qubit, dd::Qubit> replacementMap) {
+        auto targetIt = op.getTargets().begin();
+        while (targetIt != op.getTargets().end()) {
+            auto newTarget = replacementMap.find((static_cast<dd::Qubit>(*targetIt)));
+            if (newTarget != replacementMap.end()) {
+                op.setTargets(static_cast<Targets>(newTarget->second));
+            }
+            targetIt++;
+        }
+    }
+
+    void CircuitOptimizer::changeControls(Operation& op, std::map<dd::Qubit, dd::Qubit> replacementMap) {
+        auto controlIt = op.getControls().begin();
+        while (controlIt != op.getControls().end()) {
+            auto newControl = replacementMap.find((*controlIt).qubit);
+            if (newControl != replacementMap.end()) {
+                op.getControls().erase(controlIt);
+                op.getControls().insert(dd::Control{newControl->second});
+            }
+            controlIt++;
+        }
+    }
```

Comment:
Same as above. You should be able to figure it out based on the `changeTargets` suggestion above.
Small hint: A control consists of a `Qubit control` and a `Type type`. You need to copy the type of the old control to the newly emplaced one or it won't work properly.